pkg/deploy/deploy.go

```diff
@@ -208,51 +210,111 @@ func deployComponentsParallel(ctx context.Context, components []graph.SortedComp
 	return errs
 }
 
-func deployComponent(ctx context.Context, component graph.SortedComponent, clientSet ClientSet, apis api.APIs, opts DeployConfigsOptions) []error {
+type componentDeployer struct {
+	lock             sync.Mutex
+	graph            graph.ConfigGraph
+	clients          ClientSet
+	resolvedEntities entityMap
+	apis             api.APIs
+}
+
+func (c *componentDeployer) deploy(ctx context.Context) []error {
 	var errs []error
 
-	entityMap := newEntityMap(apis) //entityMap is only used to when resolving parameter references, and configs that reference each other are in the same component; no global entity map is needed
+	errChan := make(chan error)
+	for c.graph.Nodes().Len() != 0 {
+		roots := graph.Roots(c.graph)
 
-	g := component.Graph
-	sortedNodes := component.SortedNodes
+		for _, root := range roots {
+			go func(node graph2.Node) {
+				errChan <- c.deployNode(ctx, node)
+			}(root)
+		}
+
+		for range roots {
+			err := <-errChan
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
 
-	if log.Level() == loggers.LevelDebug {
-		log.WithCtxFields(ctx).Debug("Deploying configurations in current component: %v", sortedNodes)
+		// since all subroutines are done, we need not to lock here
+		for _, root := range roots {
+			c.graph.RemoveNode(root.ID())
+		}
 	}
 
-	dontDeploy := map[int64]struct{}{} //look-up map marking Nodes (by ID) that should not be deployed, because something they depend on was skipped or failed
+	close(errChan)
+
+	return errs
+}
+
+func (c *componentDeployer) deployNode(ctx context.Context, n graph2.Node) error {
+	cnf := n.(graph.ConfigNode).Config
 
-	for _, gNode := range sortedNodes {
+	ctx = context.WithValue(ctx, log.CtxKeyCoord{}, cnf.Coordinate)
+	entity, err := deployFunc(ctx, cnf, c.clients, c.apis, &c.resolvedEntities)
 
-		node := gNode.(graph.ConfigNode)
-		id := node.ID()
+	// lock changes we will make to shared variables. Writing them is trivial compared to any http request
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if err != nil {
+		failed := !errors.Is(err, skipError)
+		c.removeChildren(ctx, n, n, failed)
 
-		if _, dont := dontDeploy[id]; dont {
-			continue
+		if failed {
+			return err
 		}
+		return nil
+	}
 
-		ctx := context.WithValue(ctx, log.CtxKeyCoord{}, node.Config.Coordinate)
-		entity, err := deployFunc(ctx, node.Config, clientSet, apis, entityMap)
+	c.resolvedEntities.put(*entity)
+	log.WithCtxFields(ctx).Info("Deployment successful")
+	return nil
+}
 
-		if err != nil {
-			deploymentFailed := false
+func (c *componentDeployer) removeChildren(ctx context.Context, parent, root graph2.Node, failed bool) {
 
-			if !errors.Is(err, skipError) {
-				deploymentFailed = true
-				errs = append(errs, err)
-				if !opts.ContinueOnErr && !opts.DryRun {
-					return errs
-				}
-			}
+	parentCnf := parent.(graph.ConfigNode).Config
+	rootCnf := parent.(graph.ConfigNode).Config
```

Comment:
should this be on the `root` rather than parent?

Reply:
Nice catch!